src/main.rs

```diff
@@ -20,49 +22,62 @@ fn configure_and_try_main() -> Result<()> {
 }
 
 fn try_main(config: Config) -> Result<()> {
-    let pid_file = config.pid_file.to_owned();
-    if let Err(err) = lock(&pid_file) {
-        return Err(anyhow!(
-            "Failed to create lock file {:?}: {} (tip: use --pid-file to \
-            change the location of the lock file) ",
-            &pid_file,
-            err
-        ));
-    }
+    let mut lock_file = lock(&config)?;
+    write_pid_file(&config.pid_file)?;
 
-    let process_res = process(config);
-    if let Err(e) = unlock(&pid_file) {
-        eprint!("Failed to remove pid file {:?}: {}", pid_file, e);
-    }
+    let process_res = process(&config);
+
+    lock_file.unlock().with_context(|| "Could not remove lockfile")?;
+    remove_pid_file(&config.pid_file)?;
 
     process_res
 }
 
-pub fn lock(pid_file: &Path) -> Result<()> {
-    if pid_file.is_file() {
-        return Err(anyhow!("Lock file {:?} exists, aborting", &pid_file));
+fn lock(config: &Config) -> Result<LockFile> {
+    if !config.state_dir.exists() {
+        debug!("State directory '{:?}' does not exist yet, will try to create it.", config.state_dir);
+        std::fs::create_dir_all(&config.state_dir)
+            .with_context(|| format!("Cannot create state directory: {:?}", config.state_dir))?;
+    }
+
+    let lock_file_path = config.lock_file();
```

Comment:
As using the `fslock` crate introduces a separate lock file, maybe we should get rid of the PID file now, as the only reason for the PID file was to act as a lock.

Reply:
Or maybe it's possible to apply the lock to the PID file?

Reply:
removed pid file